File: code/Cpp/BankSystem/NormalAccount.cpp
//
// Created by 杨充 on 2026/5/25.
//

#include "NormalAccount.h"
#include <sstream>

NormalAccount::NormalAccount(const std::string &id, const std::string &name, double initBalance)
    : Account(id,name,initBalance) {
    // 调用基类构造
}

void NormalAccount::showInfo() const {
    std::cout << "[普通账户] 账号: " << accountId
     << " | 户主: " << ownerName
     << " | 余额: ¥" << balance << "\n";
}

bool NormalAccount::withdraw(double amount) {
    if (amount < 0) {
        std::cout << "[NormalAccount] 取款金额必须大于 0\n";
        return false;
    }
    if (amount > balance) {
        std::cout << "[NormalAccount] 余额不足，取款失败\n";
        return false;
    }
    balance -= amount;
    std::cout << "[NormalAccount] 取款成功，剩余 ¥" << balance << "\n";
    return true;
}

std::string NormalAccount::toCsv() const {
    std::ostringstream oss;
    oss << typeTag() << "," << accountId << "," << ownerName << "," << balance;
    return oss.str();
}

File: code/Java/BankSystem/src/com/bank/entity/NormalAccount.java
package com.bank.entity;

import com.bank.exception.InsufficientFundsException;
import com.bank.exception.InvalidAmountException;
import com.bank.persist.Persistable;

/**
 * 普通账户：不允许透支，月利率 0.1%。
 */
public class NormalAccount extends Account implements Persistable {
    private static final double MONTHLY_RATE = 0.001;   // 0.1%

    public NormalAccount(String id, String ownerName, double initialBalance) {
        super(id, ownerName, initialBalance);
    }

    @Override
    public void withdraw(double amount) {
        if (amount <= 0) throw new InvalidAmountException(amount);
        if (getBalance() < amount) {
            throw new InsufficientFundsException(amount, getBalance());
        }
        setBalanceInternal(getBalance() - amount);
    }

    @Override
    public double calcInterest() {
        return getBalance() * MONTHLY_RATE;
    }

    @Override
    public char typeTag() { return 'N'; }

    @Override
    public String toCsv() {
        return String.join(",", String.valueOf(typeTag()), getId(), getOwnerName(),
                String.valueOf(getBalance()));
    }
}
